### PR TITLE
Fix dependency specification for CuSolver variant in HiOp.

### LINKED ISSUE #571: Spack does not configure cuSOLVER LU dependency
The Spack package for HiOp variant `cusolver` sets a CMake option that is not present in HiOp anymore:
```python
self.define_from_variant("HIOP_USE_CUSOLVER", "cusolver"),
```
The correct flag is `HIOP_USE_CUSOLVER_LU`. It is used to set up KLU+cuSOLVER refactorization linear solver. Installing HiOp with Spack and `+cusolver` option does not build cuSOLVER LU module and its tests. No error message is returned. This will likely cause issues in packages like ExaGO who are using Spack to build HiOp.

To fix this, one first needs to set
```python
self.define_from_variant("HIOP_USE_CUSOLVER_LU", "cusolver"),
```
and then set Spack dependencies
```python
depends_on("cuda", when="+cusolver")
depends_on("suite-sparse", when="+cusolver")
```
I also suggest renaming this option from `cusolver` -> `cusolver_lu` to be consistent with HiOp configuration option.

This variant should not be available for HiOp versions < 0.7.1.

CC @nychiang @cnpetra @kswirydo

### PR DESCRIPTION
**Pull request by cameronrutherford**
Closes https://github.com/LLNL/hiop/issues/571.

Not sure if a conflict is the correct way to require `cuda` with `cusolver_lu`.

cc @pelesh

### CODE REVIEW

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
         description="Ultra safety checks - " "used for increased robustness and self-diagnostics",
     )
     variant("ginkgo", default=False, description="Enable/disable ginkgo solver")
-    variant("cusolver", default=False, description="Enable/disable cuSovler")
+    variant("cusolver_lu", default=False, when="@0.7.1:", description="Enable/disable cuSovler")
```
Comment (pelesh) [inline var/spack/repos/builtin/packages/hiop/package.py]: `cusolver_lu` variante enables very specific cuSolver feature, so it migth be a good idea to mention that specifically here.

```suggestion
    variant("cusolver_lu", default=False, when="@0.7.1:", description="Enable/disable cuSovler LU refactorization")

```

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 2]
```diff
@@ -97,6 +97,7 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
         depends_on("magma@{0}:".format(magma_v), when="@{0}:+rocm".format(hiop_v))
 
     depends_on("cuda@11:", when="@develop:+cuda")
+    conflicts("+cusolver_lu", when="~cuda", msg="Cusolver requires CUDA")
```
Comment (pelesh) [inline var/spack/repos/builtin/packages/hiop/package.py]: cuSolver LU refactorization needs KLU to get initial L and U factors and their sparsity patterns. 
```suggestion
    conflicts("+cusolver_lu", when="~cuda", msg="Cusolver LU refactorization requires CUDA")
    conflicts("+cusolver_lu", when="~suite-sparse", msg="Cusolver LU refactorization requires KLU from SuiteSparse library")

```

**Review by pelesh — CHANGES_REQUESTED** [review]
cuSolver LU refactorization needs a host LU solver to generate initial L and U factors. Currently, only KLU from SuiteSparse library can be used. Therefore dependency on SuiteSparse needs to be enforced in `cusolver_lu`.

**Comment by pelesh** [comment]
> Not sure if a conflict is the correct way to require `cuda` with `cusolver_lu`.

Only minor adjustments needed ;).

**Comment by pelesh** [comment]
I believe my earlier suggestion may have been misleading. Package `suite-sparse` is dependency for HiOp variants `kron` and `cusolver_lu`, but I am not sure `suite-sparse` should be a HiOp variant for itself. I am still learning Spack best practices so I will only suggest what needs to be done but let you decide on how 😄.

**Review by pelesh — APPROVED** [review]
I think this is ready to merge.

**Review by jaelynlitz — APPROVED** [review]

**Review by jaelynlitz — DISMISSED** [review]
Looks good!

**Review by pelesh — DISMISSED** [review]
Lokks good, I suggest to merge!

**Comment by cameronrutherford** [comment]
@haampie or @alalazo would so you kindly merge? You both have been very helpful in the past :)

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 3]
```diff
@@ -63,8 +63,12 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
         description="Ultra safety checks - " "used for increased robustness and self-diagnostics",
     )
     variant("ginkgo", default=False, description="Enable/disable ginkgo solver")
-    variant("cusolver", default=False, description="Enable/disable cuSovler")
-
+    variant(
+        "cusolver_lu",
+        default=False,
+        when="@0.7.1:",
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hiop/package.py]: Would this:
```suggestion
        when="+cuda @0.7.1:",
```
and removing the conflict work too?

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 4]
```diff
@@ -97,7 +101,7 @@ class Hiop(CMakePackage, CudaPackage, ROCmPackage):
         depends_on("magma@{0}:".format(magma_v), when="@{0}:+rocm".format(hiop_v))
 
     depends_on("cuda@11:", when="@develop:+cuda")
-
+    conflicts("+cusolver_lu", when="~cuda", msg="Cusolver LU refactorization requires CUDA")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hiop/package.py]:
```suggestion
```

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 3 above)
Comment (pelesh) [reply var/spack/repos/builtin/packages/hiop/package.py]: I think so. Thanks!

**Review by pelesh — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

**Pull request merged**